galsim/chromatic.py

```diff
@@ -1359,3 +1376,286 @@ def withScaledFlux(self, flux_ratio):
+
+    def withSED(self, sed='flat'):
+        """
+        Make a new InterpolatedChromaticObject with the requested SED.
+
+        @param sed    Either a galsim.SED object to assign to this object as its SED, or a string
+                      'flat' which means a flat SED (equal weight to all wavelengths).  'flat' is
+                      what should be used for PSFs, so they do not modify the flux of the objects
+                      with which they are being convolved. 
+                      [default: 'flat']
+        @return a new InterpolatedChromaticObject
+        """
+        new = self.copy()
+        if sed == 'flat':
+            new.SED = lambda w: 1.0
+        else:
+            new.SED = sed
+        return new
+
+    def evaluateAtWavelength(self, wave, force_eval=False):
+        """
+        Evaluate this InterpolatedChromaticObject at a particular wavelength, either using
+        interpolation or via direct calculation, depending on how the object was originally
+        instantiated.
+
+        @param wave     Wavelength in nanometers.
+
+        @returns the monochromatic object at the given wavelength, as an InterpolatedImage.
+        """
+        if self.waves is not None and not force_eval:
+            im, stepk, maxk = self._imageAtWavelength(wave)
+            return galsim.InterpolatedImage(im, _force_stepk=stepk, _force_maxk=maxk)
+        else:
+            return self.simpleEvaluateAtWavelength(wave)
+
+    def _imageAtWavelength(self, wave):
+        """
+        Get an image of the object at a particular wavelength, using linear interpolation between
+        the originally-stored images.
+        """
+        # First, some wavelength-related sanity checks.
+        if self.waves is None:
+            raise RuntimeError("Requested image at some wavelength when doing direct calculation!")
+        if wave < min(self.waves) or wave > max(self.waves):
+            raise RuntimeError("Requested wavelength %.1f is outside the allowed range:"
+                               " %.1f to %.1f nm"%(wave, min(self.waves), max(self.waves)))
+
+        # Figure out where the supplied wavelength is compared to the list of wavelengths on which
+        # images were originally tabulated.
+        lower_idx = np.searchsorted(self.waves, wave)-1
+        # There can be edge issues, so watch out for that:
+        if lower_idx < 0: lower_idx = 0
+        if lower_idx == len(self.waves)-1: lower_idx = len(self.waves)-1
+        frac = (wave-self.waves[lower_idx]) / (self.waves[lower_idx+1]-self.waves[lower_idx])
+
+        # Actually do the interpolation for the image, stepK, and maxK.
+        im = _linearInterp(self.ims, frac, lower_idx)
+        stepk = _linearInterp(self.stepK_vals, frac, lower_idx)
+        maxk = _linearInterp(self.maxK_vals, frac, lower_idx)
+        return self.SED(wave)*im, stepk, maxk
+
+    def drawImage(self, bandpass, force_eval=False, image=None, **kwargs):
```

Comment:
I'm not sure if it would make a big timing difference, but I think this could be rewritten to use fewer sums.  For instance, if len(self.waves)=10, then the final profile will be a linear combination of no more than 10 terms, regardless of the value of len(self.wave_list) or len(bandpass.wave_list).  This is a consequence of the linear interpolation; a more generic wavelength-dependence doesn't have this property (although other polynomial interpolation rules like Simpson's rule probably do).  

It might be worth looping through the elements of the combined wave list here to build up the coefficient of each term and then forming the linear combination instead of using DirectSampleIntegrator which doesn't know about the linear interpolation.

Likewise, I don't think you need to check the values of stepk and maxk at each wavelength in the combined wave list; it's probably sufficient to check at the elements of self.waves.